File: ios/NorthAx/Engine/PlanMatchingEngine.swift
import SwiftUI

/// Completion state of a planned session, derived by matching it against
/// workouts imported from intervals.icu / Garmin / Apple Health.
enum SessionCompletion {
    case planned   // scheduled today or in the future, not yet done
    case done      // a matching imported workout was found
    case missed    // the planned day has passed with no matching workout
    case rest      // no session scheduled

    var label: String {
        switch self {
        case .planned: return "Planned"
        case .done:    return "Done"
        case .missed:  return "Missed"
        case .rest:    return "Rest"
        }
    }

    var color: Color {
        switch self {
        case .planned: return .axAccent
        case .done:    return .axGreen
        case .missed:  return .axRed
        case .rest:    return .axTertiary
        }
    }

    var icon: String {
        switch self {
        case .planned: return "circle"
        case .done:    return "checkmark.circle.fill"
        case .missed:  return "xmark.circle"
        case .rest:    return "moon"
        }
    }
}

/// A planned session paired with its completion state and (when done) the
/// imported workout it matched.
struct SessionMatch: Identifiable {
    let id: UUID
    let day: PlannedDay
    let session: PlannedSession
    let completion: SessionCompletion
    let activity: GarminActivity?

    /// Stable key for caching daily switch suggestions (§9) — survives plan
    /// reloads (unlike the random `session.id`) as long as the plan is unchanged.
    var suggestionKey: String { SessionMatch.suggestionKey(day: day, session: session) }

    static func suggestionKey(day: PlannedDay, session: PlannedSession) -> String {
        "\(dateKey.string(from: day.date))|\(session.domain.rawValue)|\(session.title)"
    }

    private static let dateKey: DateFormatter = {
        let f = DateFormatter(); f.dateFormat = "yyyy-MM-dd"; return f
    }()
}

/// One navigable week (§11): the plan (or a synthesized past week) plus its
/// matches. `isHistorical` weeks are built from imported activities only.
struct WeekData {
    let offset: Int
    let week: WeeklyPlan
    let matches: [SessionMatch]
    let isHistorical: Bool
}

/// Client-side matching of a week's planned sessions to imported workouts
/// (§7 of the plan). Match on same calendar day + same sport; when several
/// workouts fit, pick the one closest in duration to the planned session.
enum PlanMatchingEngine {
    static func matches(week: WeeklyPlan, activities: [GarminActivity], today: Date = Date()) -> [SessionMatch] {
        let cal = Calendar.current
        let startToday = cal.startOfDay(for: today)
        var out: [SessionMatch] = []
        var matchedActivityIDs: Set<GarminActivity.ID> = []
        for day in week.days where !day.isRest && !day.sessions.isEmpty {
            for session in day.sessions {
                let sameDaySameSport = activities.filter {
                    cal.isDate($0.startTime, inSameDayAs: day.date) && $0.type.domain == session.domain
                        && !matchedActivityIDs.contains($0.id)
                }
                let matched = sameDaySameSport.min {
                    abs($0.duration / 60 - Double(session.duration))
                        < abs($1.duration / 60 - Double(session.duration))
                }
                if let matched { matchedActivityIDs.insert(matched.id) }
                let completion: SessionCompletion
                if matched != nil {
                    completion = .done
                } else if cal.startOfDay(for: day.date) < startToday {
                    completion = .missed
                } else {
                    completion = .planned
                }
                out.append(SessionMatch(id: session.id, day: day, session: session,
                                        completion: completion, activity: matched))
            }
        }

        // Surface imported workouts that don't correspond to any planned session
        // (unplanned / extra sessions) as done entries, so a workout the athlete
        // did off-plan still shows up on the week instead of vanishing (§7).
        for day in week.days {
            let extras = activities.filter {
                cal.isDate($0.startTime, inSameDayAs: day.date) && !matchedActivityIDs.contains($0.id)
            }
            for a in extras {
                matchedActivityIDs.insert(a.id)
                let session = PlannedSession(
                    domain: a.type.domain, title: a.name, subtitle: "",
                    duration: Int(a.duration / 60), intensityLabel: "Completed"
                )
                out.append(SessionMatch(id: session.id, day: day, session: session,
                                        completion: .done, activity: a))
            }
        }

        // Chronological order regardless of planned/unplanned. Stable within a day
        // (original index tiebreak) so planned sessions stay ahead of extras.
        return out.enumerated()
            .sorted { ($0.element.day.date, $0.offset) < ($1.element.day.date, $1.offset) }
            .map(\.element)
    }

    /// Roll a day's session matches up to a single state for the week strip:
    /// rest → any missed → all done → otherwise planned.
    static func dayState(day: PlannedDay, matches: [SessionMatch]) -> SessionCompletion {
        let dayMatches = matches.filter { $0.day.date == day.date }
        if day.isRest || day.sessions.isEmpty {
            // A rest day still marks done if an unplanned workout was imported for it.
            return dayMatches.contains { $0.completion == .done } ? .done : .rest
        }
        if dayMatches.isEmpty { return .planned }
        if dayMatches.contains(where: { $0.completion == .missed }) { return .missed }
        if dayMatches.allSatisfy({ $0.completion == .done }) { return .done }
        return .planned
    }
}

File: android/app/src/main/java/app/northax/domain/engine/PlanMatchingEngine.kt
package app.northax.domain.engine

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AddCircle
import androidx.compose.material.icons.filled.Cancel
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.NightsStay
import androidx.compose.material.icons.outlined.Circle
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import app.northax.domain.model.GarminActivity
import app.northax.domain.model.PlannedDay
import app.northax.domain.model.PlannedSession
import app.northax.domain.model.WeeklyPlan
import app.northax.ui.theme.Ax
import java.time.LocalDate
import java.time.ZoneId
import kotlin.math.abs

/**
 * Completion state of a planned session, derived by matching it against
 * workouts imported from intervals.icu / Garmin / Strava.
 */
enum class SessionCompletion {
    Planned, // scheduled today or in the future, not yet done
    Done,    // a matching imported workout was found
    Extra,   // an imported workout with no planned session (off-plan)
    Missed,  // the planned day has passed with no matching workout
    Rest;    // no session scheduled

    val label: String
        get() = when (this) {
            Planned -> "Planned"; Done -> "Done"; Extra -> "Extra"; Missed -> "Missed"; Rest -> "Rest"
        }

    val color: Color
        get() = when (this) {
            Planned -> Ax.Accent
            Done -> Ax.Green
            Extra -> Ax.Purple
            Missed -> Ax.Red
            Rest -> Ax.Tertiary
        }

    val icon: ImageVector
        get() = when (this) {
            Planned -> Icons.Outlined.Circle
            Done -> Icons.Filled.CheckCircle
            Extra -> Icons.Filled.AddCircle
            Missed -> Icons.Filled.Cancel
            Rest -> Icons.Filled.NightsStay
        }

    /** A workout actually happened — planned (Done) or off-plan (Extra). */
    val isCompleted: Boolean get() = this == Done || this == Extra
}

/**
 * A planned session paired with its completion state and (when done) the
 * imported workout it matched.
 */
data class SessionMatch(
    val id: String,
    val day: PlannedDay,
    val session: PlannedSession,
    val completion: SessionCompletion,
    val activity: GarminActivity?,
) {
    /** Stable key for caching daily switch suggestions — survives plan reloads. */
    val suggestionKey: String get() = suggestionKey(day, session)

    companion object {
        fun suggestionKey(day: PlannedDay, session: PlannedSession): String =
            "${day.date}|${session.domain.raw}|${session.title}"
    }
}

/** One navigable week: the plan (or a synthesized past week) plus its matches. */
data class WeekData(
    val offset: Int,
    val week: WeeklyPlan,
    val matches: List<SessionMatch>,
    val isHistorical: Boolean,
)

/**
 * Client-side matching of a week's planned sessions to imported workouts.
 * Match on same calendar day + same sport; when several workouts fit, pick the
 * one closest in duration to the planned session.
 */
object PlanMatchingEngine {

    private fun GarminActivity.localDate(): LocalDate =
        startTime.atZone(ZoneId.systemDefault()).toLocalDate()

    fun matches(
        week: WeeklyPlan,
        activities: List<GarminActivity>,
        today: LocalDate = LocalDate.now(),
    ): List<SessionMatch> {
        val out = mutableListOf<SessionMatch>()
        val matchedActivityIds = mutableSetOf<String>()

        for (day in week.days) {
            if (day.isRest || day.sessions.isEmpty()) continue
            for (session in day.sessions) {
                val sameDaySameSport = activities.filter {
                    it.localDate() == day.date && it.type.domain == session.domain &&
                        it.id !in matchedActivityIds
                }
                val matched = sameDaySameSport.minByOrNull {
                    abs(it.durationSeconds / 60.0 - session.duration)
                }
                if (matched != null) matchedActivityIds.add(matched.id)
                val completion = when {
                    matched != null -> SessionCompletion.Done
                    day.date.isBefore(today) -> SessionCompletion.Missed
                    else -> SessionCompletion.Planned
                }
                out.add(SessionMatch(session.id, day, session, completion, matched))
            }
        }

        // Surface imported workouts that don't correspond to any planned session
        // as extra entries, so an off-plan workout still shows up on the week.
        for (day in week.days) {
            val extras = activities.filter {
                it.localDate() == day.date && it.id !in matchedActivityIds
            }
            for (a in extras) {
                matchedActivityIds.add(a.id)
                out.add(extraMatch(a, day))
            }
        }

        // Chronological order regardless of planned/unplanned. Stable within a
        // day (original index tiebreak) so planned sessions stay ahead of extras.
        return out.withIndex()
            .sortedWith(compareBy({ it.value.day.date }, { it.index }))
            .map { it.value }
    }

    /**
     * Today's rows for the dashboard: today's matches from the current week,
     * plus an extra entry for any workout done today that no plan day covers —
     * a lapsed or missing plan must not hide what the athlete actually did.
     */
    fun todayMatches(
        week: WeeklyPlan?,
        activities: List<GarminActivity>,
        today: LocalDate = LocalDate.now(),
    ): List<SessionMatch> {
        val weekMatches = week?.let { matches(it, activities, today) } ?: emptyList()
        val out = weekMatches.filter { it.day.date == today }.toMutableList()
        val seen = out.mapNotNull { it.activity?.id }.toSet()
        val day = PlannedDay(date = today, sessions = emptyList(), isRest = false)
        for (a in activities) {
            if (a.localDate() != today || a.id in seen) continue
            out.add(extraMatch(a, day))
        }
        return out
    }

    /** Off-plan workout as a match row. The id derives from the activity so it
     *  stays stable across recomputations (extras have no persisted session). */
    private fun extraMatch(a: GarminActivity, day: PlannedDay): SessionMatch {
        val session = PlannedSession(
            id = "extra-${a.id}",
            domain = a.type.domain, title = a.name, subtitle = "",
            duration = (a.durationSeconds / 60).toInt(), intensityLabel = "Completed",
        )
        return SessionMatch(session.id, day, session, SessionCompletion.Extra, a)
    }

    /**
     * Roll a day's session matches up to a single state for the week strip:
     * rest → any missed → all done → otherwise planned.
     */
    fun dayState(day: PlannedDay, matches: List<SessionMatch>): SessionCompletion {
        val dayMatches = matches.filter { it.day.date == day.date }
        if (day.isRest || day.sessions.isEmpty()) {
            // A rest day still marks done if an unplanned workout was imported for it.
            return if (dayMatches.any { it.completion.isCompleted }) SessionCompletion.Done
            else SessionCompletion.Rest
        }
        if (dayMatches.isEmpty()) return SessionCompletion.Planned
        if (dayMatches.any { it.completion == SessionCompletion.Missed }) return SessionCompletion.Missed
        if (dayMatches.all { it.completion.isCompleted }) return SessionCompletion.Done
        return SessionCompletion.Planned
    }
}
